apply_DDT.py

```diff
@@ -0,0 +1,574 @@
+#==============================================================================
+# apply_DDT.py ----------------------------------------------------------------
+#------------------------------------------------------------------------------
+# Author(s): Brendan Regnery, Sara Nabili -------------------------------------
+#------------------------------------------------------------------------------
+# Applies a DDT to a trained BDT model ----------------------------------------
+#    (Designed Decorrelated Tagger, https://arxiv.org/pdf/1603.00027.pdf) -----
+#------------------------------------------------------------------------------
+
+import os, os.path as osp, glob, pickle, logging, warnings, json, math, re
+from time import strftime
+from collections import OrderedDict
+from common import create_DDT_map_dict, calculate_varDDT
+import json
+import argparse
+
+import numpy as np
+import matplotlib
+matplotlib.use('Agg') # prevents opening displays (fast), must use before pyplot
+import matplotlib.pyplot as plt
+from sklearn.metrics import confusion_matrix, roc_curve, roc_auc_score, auc
+import pandas as pd
+import xgboost as xgb
+from scipy.ndimage import gaussian_filter
+import mplhep as hep
+hep.style.use("CMS") # CMS plot style
+import svj_ntuple_processing as svj
+
+#disable tex formating while having problems
+#plt.rcParams['text.usetex'] = False
+
+np.random.seed(1001)
+
+from common import read_traing_features, logger, DATADIR, Columns, time_and_log, imgcat, set_mpl_fontsize, columns_to_numpy
+
+#------------------------------------------------------------------------------
+# Global variables and user input arguments -----------------------------------
+#------------------------------------------------------------------------------
+
+training_features = []
+
+def parse_arguments():
+    parser = argparse.ArgumentParser(description='Process some inputs.')
+
+    parser.add_argument('--bkg-files', default='data/bkg_20240718/Summer20UL*/QCD*.npz', help='Background data files')
+    parser.add_argument('--sig-files', default='data/sig_20240718/sig_*/*.npz', help='Signal data files')
+
+    # BDT and ddt model
+    parser.add_argument('--bdt_file', default='models/svjbdt_obj_rev_version.json', help='BDT model file')
+    parser.add_argument('--ddt_map_file', default='models/ddt_obj_rev_version.json', help='DDT map file')
+
+    parser.add_argument('--lumi', type=float, default=137600, help='Luminosity')
+
+    parser.add_argument('--sig_bdt_cut', type=float, default=0.65, help='BDT cut for signal plotting')
+
+    # Choose the BDT cut values that you want to make for the DDT
+    # or that are in the DDT you are loading
```

Comment:
I spent some time thinking about this point.

The optimized value is 0.65, which I set as default. The main issue is that I was trying to keep the code adaptable. The code can calculate the DDT at various BDT values or upload a map that has all the values saved for various BDT values. The user may decide to make a new map, with finer bdt scoring such as [... 0.645 0.65 0.655 ...]. 

My way of getting around this issue is by having a default value. That can be set to the most recent 'best value'

If you have a better idea let me know. I just don't want to redo the calculations every time... because the code could simply apply the DDT and make some diagnostic plots.

Reply:
Additionally added a comment by the argument